kotlintest-assertions-arrow/src/main/kotlin/io/kotlintest/assertions/arrow/gen/gen.kt

```diff
@@ -1,15 +1,76 @@
-//package io.kotlintest.assertions.arrow.gen
-//
-//import arrow.core.Tuple2
-//import io.kotlintest.properties.Gen
-//
-//fun <A, B, T> genT(fn: Tuple2<A, B>): Gen<T> = object : Gen<T> {
-//
-//  init {
-//    println(fn::class.typeParameters)
-//  }
-//
-//  override fun constants(): Iterable<T> = emptyList()
-//  override fun random(): Sequence<T> =
-//      generateSequence { TODO() }
-//}
\ No newline at end of file
+package io.kotlintest.assertions.arrow.gen
+
+import arrow.Kind
+import arrow.core.Either
+import arrow.extension
+import arrow.typeclasses.*
+import io.kotlintest.properties.ForGen
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.GenOf
+import io.kotlintest.properties.fix
+
+@extension
```

Comment:
What does @extension do

Reply:
It generates top level extension functions for all the helper functions defined in the typeclass. You give one function, like `map` for `Functor`, and you get the others for free: https://arrow-kt.io/docs/apidocs/arrow-typeclasses/arrow.typeclasses/-functor/index.html#functions

Reply:
Because the other functions can be implemented using map, you only need to provide map and the rest can be derived? Why do you need @extension and not just a regular abstract method with other definitions in the interface/typeclass ?

Reply:
We did that for a while. For those functions to be available you needed to do:

```
Option.functor().run { myGen.tupleLeft(1) }
```

So at some point we put all typeclasses together and made it a bit nicer:

```
ForOption {
  myGen.tupleLeft(1)
}
```

until we finally settled with creating top-level extension functions.